File: src/mods/physics.cpp
#include "physics.h"
#include "mkb/mkb.h"
#include "mods/freecam.h"
#include "systems/pad.h"
#include "systems/pref.h"
#include "utils/draw.h"
#include "utils/macro_utils.h"

namespace physics {

static f32 s_orig_friction;     // = 0.010
static f32 s_orig_restitution;  // = 0.50

bool using_custom_physics() {
    return pref::get(pref::U8Pref::PhysicsPreset);
}

void init() {
    s_orig_friction = mkb::ball_friction;
    s_orig_restitution = mkb::ball_restitution;
}

static void change_physics() {
    // restore physics momentarily
    mkb::ball_friction = s_orig_friction;
    mkb::ball_restitution = s_orig_restitution;
    mkb::balls[mkb::curr_player_idx].restitution = s_orig_restitution;

    // update physics depending on preset
    switch (PhysicsPreset(pref::get(pref::U8Pref::PhysicsPreset))) {
        case PhysicsPreset::Default: {
            break;
        }
        case PhysicsPreset::LightBall: {
            bool paused_now = *reinterpret_cast<u32*>(0x805BC474) & 8;
            if (mkb::sub_mode == mkb::SMD_GAME_PLAY_MAIN && !paused_now) {
                mkb::balls[mkb::curr_player_idx].vel.y += 0.005f;
            }
            break;
        }
        case PhysicsPreset::NoFriction: {
            mkb::ball_friction = 0.0f;
            break;
        }
        case PhysicsPreset::HeavyBall: {
            bool paused_now = *reinterpret_cast<u32*>(0x805BC474) & 8;
            if (mkb::sub_mode == mkb::SMD_GAME_PLAY_MAIN && !paused_now) {
                mkb::balls[mkb::curr_player_idx].vel.y -= 0.005f;
            }
            break;
        }
        case PhysicsPreset::BouncyBall: {
            mkb::ball_restitution = 1.20f;
            mkb::balls[mkb::curr_player_idx].restitution = 1.20f;
            break;
        }
        case PhysicsPreset::StickyBall: {
            mkb::ball_restitution = 0.01f;
            mkb::balls[mkb::curr_player_idx].restitution = 0.01f;
            break;
        }
        case PhysicsPreset::HighFriction: {
            mkb::ball_friction = 0.020f;
            break;
        }
        case PhysicsPreset::JumpPhysics: {
            mkb::ball_friction = 0.015f;
            mkb::ball_restitution = 0.25f;
            mkb::balls[mkb::curr_player_idx].restitution = 0.25f;
            break;
        }
    }
}

void tick() {
    change_physics();
}

void disp() {
    if (mkb::sub_mode != mkb::SMD_GAME_READY_INIT && mkb::sub_mode != mkb::SMD_GAME_READY_MAIN &&
        mkb::sub_mode != mkb::SMD_GAME_PLAY_INIT && mkb::sub_mode != mkb::SMD_GAME_PLAY_MAIN)
        return;

    if (using_custom_physics() && pref::get(pref::BoolPref::CustomPhysicsDisp) &&
        !freecam::should_hide_hud()) {
        mkb::textdraw_reset();
        mkb::textdraw_set_font(mkb::FONT32_ASC_8x16);
        u32 x = 634;
        u32 y = 474;
        mkb::textdraw_set_pos(x, y);
        mkb::textdraw_set_alignment(mkb::ALIGN_UPPER_LEFT);
        mkb::textdraw_set_scale(1.2, 0.9);
        mkb::GXColor color = draw::WHITE;
        mkb::textdraw_set_mul_color(RGBA(color.r, color.g, color.b, color.a));
        mkb::textdraw_print("Custom Physics");
    }
}

}  // namespace physics

File: src/mods/physics.rs
#![cfg(feature = "mkb2")]

use num_enum::TryFromPrimitive;

use crate::{
    app::AppContext,
    mkb2::mkb2,
    systems::{
        draw,
        pref::{BoolPref, Pref, U8Pref},
    },
};

use super::freecam::Freecam;

struct Context<'a> {
    pref: &'a mut Pref,
    freecam: &'a mut Freecam,
}

#[derive(Copy, Clone, PartialEq, Eq, TryFromPrimitive)]
#[repr(u8)]
pub enum PhysicsPreset {
    Default,
    LightBall,
    NoFriction,
    HeavyBall,
    BouncyBall,
    StickyBall,
    HighFriction,
    JumpPhysics,
}

pub struct Physics {
    orig_friction: f32,    // = 0.010
    orig_restitution: f32, // = 0.50
}

impl Default for Physics {
    fn default() -> Self {
        unsafe {
            Self {
                orig_friction: mkb2::ball_friction,
                orig_restitution: mkb2::ball_restitution,
            }
        }
    }
}

impl Physics {
    pub fn using_custom_physics(&self, pref: &Pref) -> bool {
        let preset: PhysicsPreset = pref.get_u8(U8Pref::PhysicsPreset).try_into().unwrap();
        preset != PhysicsPreset::Default
    }

    fn change_physics(&mut self, cx: &mut Context) {
        unsafe {
            // restore physics momentarily
            mkb2::ball_friction = self.orig_friction;
            mkb2::ball_restitution = self.orig_restitution;
            mkb2::balls[mkb2::curr_player_idx as usize].restitution = self.orig_restitution;

            // update physics depending on preset
            let preset = cx.pref.get_u8(U8Pref::PhysicsPreset).try_into().unwrap();
            match preset {
                PhysicsPreset::Default => {}
                PhysicsPreset::LightBall => {
                    let paused_now = *(0x805BC474 as *const u32) & 8 != 0;
                    if mkb2::sub_mode == mkb2::SMD_GAME_PLAY_MAIN && !paused_now {
                        mkb2::balls[mkb2::curr_player_idx as usize].vel.y += 0.005;
                    }
                }
                PhysicsPreset::NoFriction => {
                    mkb2::ball_friction = 0.0;
                }
                PhysicsPreset::HeavyBall => {
                    let paused_now = *(0x805BC474 as *const u32) & 8 != 0;
                    if mkb2::sub_mode == mkb2::SMD_GAME_PLAY_MAIN && !paused_now {
                        mkb2::balls[mkb2::curr_player_idx as usize].vel.y -= 0.005;
                    }
                }
                PhysicsPreset::BouncyBall => {
                    mkb2::ball_restitution = 1.20;
                    mkb2::balls[mkb2::curr_player_idx as usize].restitution = 1.20;
                }
                PhysicsPreset::StickyBall => {
                    mkb2::ball_restitution = 0.01;
                    mkb2::balls[mkb2::curr_player_idx as usize].restitution = 0.01;
                }
                PhysicsPreset::HighFriction => {
                    mkb2::ball_friction = 0.020;
                }
                PhysicsPreset::JumpPhysics => {
                    mkb2::ball_friction = 0.015;
                    mkb2::ball_restitution = 0.25;
                    mkb2::balls[mkb2::curr_player_idx as usize].restitution = 0.25;
                }
            }
        }
    }

    pub fn tick(&mut self, cx: &AppContext) {
        let cx = &mut Context {
            pref: &mut cx.pref.borrow_mut(),
            freecam: &mut cx.freecam.borrow_mut(),
        };
        self.change_physics(cx);
    }

    pub fn draw(&self, cx: &AppContext) {
        let cx = &mut Context {
            pref: &mut cx.pref.borrow_mut(),
            freecam: &mut cx.freecam.borrow_mut(),
        };
        unsafe {
            if mkb2::sub_mode != mkb2::SMD_GAME_READY_INIT
                && mkb2::sub_mode != mkb2::SMD_GAME_READY_MAIN
                && mkb2::sub_mode != mkb2::SMD_GAME_PLAY_INIT
                && mkb2::sub_mode != mkb2::SMD_GAME_PLAY_MAIN
            {
                return;
            }

            if self.using_custom_physics(cx.pref)
                && cx.pref.get_bool(BoolPref::CustomPhysicsDisp)
                && !cx.freecam.should_hide_hud(cx.pref)
            {
                mkb2::textdraw_reset();
                mkb2::textdraw_set_font(mkb2::FONT32_ASC_8x16);
                let x = 634.0;
                let y = 474.0;
                mkb2::textdraw_set_pos(x, y);
                mkb2::textdraw_set_alignment(mkb2::ALIGN_UPPER_LEFT as mkb2::SpriteAlignment);
                mkb2::textdraw_set_scale(1.2, 0.9);
                let color = draw::WHITE;
                mkb2::textdraw_set_mul_color(color.into());
                mkb2::textdraw_print(c"Custom Physics".as_ptr() as *mut _);
            }
        }
    }
}
